pkg/recorder/recorder.go

```diff
@@ -0,0 +1,632 @@
+}
+
+func (r *Recorder) queueAddDatapathFilter(ri *RecInfo, i int) {
+	if r.queue.ri == nil {
+		r.queue.ri = ri
+	}
+	r.queue.add = append(r.queue.add, &ri.Filters[i])
+}
+
+func (r *Recorder) queueDelDatapathFilter(ri *RecInfo, i int) {
+	if r.queue.ri == nil {
+		r.queue.ri = ri
+	}
+	filter := &ri.Filters[i]
+	hash := hashTuple(filter)
+	// If the recorder updated an existing filter element which sits
+	// in both queues, then we do not need any change in the BPF data
+	// path, and can avoid temporary recorder disruption. Hence, add/del
+	// queues strictly only ever contain entries that need change.
+	for i, e := range r.queue.add {
+		if hashTuple(e) == hash {
+			r.queue.add = queuePurge(r.queue.add, i)
+			return
+		}
+	}
+	r.queue.del = append(r.queue.del, filter)
+}
+
+func (r *Recorder) deleteRecInfoLocked(ri *RecInfo, withID bool) bool {
+	triggerRegen := false
+	for i, filter := range ri.Filters {
+		mask := convertTupleToMask(filter)
+		maskHash := hashMask(&mask)
+		if rm, found := r.recMask[maskHash]; found {
+			rm.users--
+			if rm.users == 0 {
+				delete(r.recMask, maskHash)
+				triggerRegen = true
+			}
+		}
+		r.queueDelDatapathFilter(ri, i)
+	}
+	if withID {
+		delete(r.recByID, ri.ID)
+	}
+	return triggerRegen
+}
+
+func (r *Recorder) DeleteRecorder(id ID) (bool, error) {
+	r.Lock()
+	defer r.Unlock()
+	if recInfo, found := r.recByID[id]; found {
+		return true, r.applyDatapath(r.deleteRecInfoLocked(recInfo, true))
+	}
+	return false, nil
+}
+
+func (r *Recorder) createRecInfoLocked(ri *RecInfo, withID bool) bool {
+	if withID {
+		r.recByID[ri.ID] = ri
+	}
+	triggerRegen := false
+	for i, filter := range ri.Filters {
+		mask := convertTupleToMask(filter)
+		maskHash := hashMask(&mask)
+		if rm, found := r.recMask[maskHash]; found {
+			rm.users++
+		} else {
+			ones := countMaskOnes(mask)
+			rm := &RecMask{
+				users: 1,
+				mask:  mask,
+				prio:  ones,
+			}
+			r.recMask[maskHash] = rm
+			triggerRegen = true
+		}
+		r.queueAddDatapathFilter(ri, i)
+	}
+	return triggerRegen
+}
+
+func (r *Recorder) updateRecInfoLocked(riNew, riOld *RecInfo) error {
+	triggerRegen := false
+	if r.createRecInfoLocked(riNew, true) {
+		triggerRegen = true
+	}
+	if r.deleteRecInfoLocked(riOld, false) {
+		triggerRegen = true
+	}
+	return r.applyDatapath(triggerRegen)
+}
+
+func deepCopyPrefix(p net.IPNet) net.IPNet {
+	out := net.IPNet{
+		IP:   make([]byte, len(p.IP)),
+		Mask: make([]byte, len(p.Mask)),
+	}
+	copy(out.IP, p.IP)
+	copy(out.Mask, p.Mask)
+	return out
+}
+
+func deepCopyRecInfo(recInfo *RecInfo) *RecInfo {
+	ri := &RecInfo{
+		ID:      recInfo.ID,
+		CapLen:  recInfo.CapLen,
+		Filters: []RecorderTuple{},
+	}
+	for _, filter := range recInfo.Filters {
+		f := RecorderTuple{
+			SrcPort: filter.SrcPort,
+			DstPort: filter.DstPort,
+			Proto:   filter.Proto,
+		}
+		f.SrcPrefix = deepCopyPrefix(filter.SrcPrefix)
+		f.DstPrefix = deepCopyPrefix(filter.DstPrefix)
+		ri.Filters = append(ri.Filters, f)
+	}
+	return ri
+}
```

Comment:
Please use the deepcopy generators that we have instead of doing it manually.

Reply:
Where do I find these, do you have some pointers?

Reply:
here:
https://github.com/cilium/cilium/blob/6483724330b59947dce0c400660561f410031830/pkg/bpf/doc.go
https://github.com/cilium/cilium/blob/070214f089e78bf047ce671bd15d762174654ab6/pkg/bpf/endpoint.go#L34
https://github.com/cilium/cilium/blob/ae23400cb944c722f629b9b3f0d7e8af58af1b34/Makefile#L427